# LLM Agents & Deep Q-Learning with Atari Games


- DQN Agent for DemonAttack-v5  
- Author: Anusree Mohanan  
- Partially adapted from: Stable-Baselines3 official DQN tutorial  
- Source: https://stable-baselines3.readthedocs.io/en/master/modules/dqn.html  
- License: MIT (SB3)


# Installations and Imports

In [5]:
%pip install gym

Note: you may need to restart the kernel to use updated packages.


In [6]:
%pip install opencv-python

Note: you may need to restart the kernel to use updated packages.


In [7]:
%pip install "gymnasium[atari,accept-rom-license]"

  Using cached AutoROM-0.4.2-py3-none-any.whl.metadata (2.8 kB)
  Using cached AutoROM.accept-rom-license-0.6.1.tar.gz (434 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Using cached ale_py-0.8.1-cp311-cp311-macosx_11_0_arm64.whl.metadata (8.1 kB)
Using cached AutoROM-0.4.2-py3-none-any.whl (16 kB)
Using cached ale_py-0.8.1-cp311-cp311-macosx_11_0_arm64.whl (1.0 MB)
  Created wheel for AutoROM.accept-rom-license: filename=autorom_accept_rom_license-0.6.1-py3-none-any.whl size=446709 sha256=d19a0efc77b0077de50cae222aa114ff17a4c449aef7b1dfdd6c0fc221c65528
  Stored in directory: /Users/anusreemohanan/Library/Caches/pip/wheels/bc/fc/c6/8aa657c0d2089982f2dabd110efc68c61eb49831fdb7397351
Successfully built AutoROM.accept-rom-license
  Attempting uninstall: ale-py
    Found existing installation: ale-py 0.11.2
    Uninstalling ale-py-0.11.2:
      Successfully uninstalled ale-py-0.11.2
  Attempting 

In [10]:
import numpy as np
import cv2
from stable_baselines3 import DQN
import gym
from stable_baselines3.common.vec_env import DummyVecEnv

In [11]:
import gymnasium as gym
from stable_baselines3.common.env_util import make_atari_env
from stable_baselines3.common.vec_env import VecFrameStack

In [ ]:
%pip uninstall gymnasium ale-py -y
%pip install "gymnasium[atari,accept-rom-license]"

Found existing installation: gymnasium 1.1.1
Uninstalling gymnasium-1.1.1:
  Successfully uninstalled gymnasium-1.1.1
Found existing installation: ale-py 0.11.2
Uninstalling ale-py-0.11.2:
  Successfully uninstalled ale-py-0.11.2
Note: you may need to restart the kernel to use updated packages.
  Using cached gymnasium-1.2.0-py3-none-any.whl.metadata (9.9 kB)
  Using cached ale_py-0.11.2-cp311-cp311-macosx_13_0_arm64.whl.metadata (9.0 kB)
Using cached gymnasium-1.2.0-py3-none-any.whl (944 kB)
Using cached ale_py-0.11.2-cp311-cp311-macosx_13_0_arm64.whl (2.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [ale_py]
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
stable-baselines3 2.6.0 requires gymnasium<1.2.0,>=0.29.1, but you have gymnasium 1.2.0 which is incompatible.
Note: you may need to restart the kernel to use updated packages.


In [ ]:
%pip uninstall gymnasium ale-py -y
%pip install "gymnasium[atari]==0.29.1" ale-py==0.8.1

Found existing installation: gymnasium 1.2.0
Uninstalling gymnasium-1.2.0:
  Successfully uninstalled gymnasium-1.2.0
Found existing installation: ale-py 0.11.2
Uninstalling ale-py-0.11.2:
  Successfully uninstalled ale-py-0.11.2
Note: you may need to restart the kernel to use updated packages.
  Using cached ale_py-0.8.1-cp311-cp311-macosx_11_0_arm64.whl.metadata (8.1 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 953.9/953.9 kB 3.7 MB/s eta 0:00:00a 0:00:01
Using cached ale_py-0.8.1-cp311-cp311-macosx_11_0_arm64.whl (1.0 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3/3 [shimmy]2m1/3 [ale-py]
Note: you may need to restart the kernel to use updated packages.


In [3]:
import gymnasium as gym
print("ALE/DemonAttack-v5" in gym.envs.registry)

True


In [ ]:
%pip install tensorboard

  Using cached absl_py-2.3.1-py3-none-any.whl.metadata (3.3 kB)
  Using cached markdown-3.8.2-py3-none-any.whl.metadata (5.1 kB)
  Using cached tensorboard_data_server-0.7.2-py3-none-any.whl.metadata (1.1 kB)
  Using cached werkzeug-3.1.3-py3-none-any.whl.metadata (3.7 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 3.4 MB/s eta 0:00:00a 0:00:01
Using cached tensorboard_data_server-0.7.2-py3-none-any.whl (2.4 kB)
Using cached absl_py-2.3.1-py3-none-any.whl (135 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.6/10.6 MB 3.7 MB/s eta 0:00:0000:010:01
Using cached markdown-3.8.2-py3-none-any.whl (106 kB)
Using cached werkzeug-3.1.3-py3-none-any.whl (224 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7/7 [tensorboard] [tensorboard]
Note: you may need to restart the kernel to use updated packages.


### Environment Setup

The Atari environment **DemonAttack-v5** was initialized using `make_atari_env()` from Stable-Baselines3. To capture temporal dynamics, a stack of 4 consecutive frames was created using `VecFrameStack`. This preprocessing setup allows the agent to learn motion patterns effectively, which is essential in Atari-style environments.


In [12]:
from stable_baselines3 import DQN
from stable_baselines3.common.atari_wrappers import AtariWrapper
from stable_baselines3.common.vec_env import VecFrameStack, DummyVecEnv
from stable_baselines3.common.env_util import make_atari_env

# Create Atari environment
env = make_atari_env("ALE/DemonAttack-v5", n_envs=1, seed=0)
env = VecFrameStack(env, n_stack=4)

In [10]:
%pip install moviepy

  Using cached tqdm-4.67.1-py3-none-any.whl.metadata (57 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 5.2 MB/s eta 0:00:00a 0:00:01
Using cached tqdm-4.67.1-py3-none-any.whl (78 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6/6 [moviepy]m2/6 [imageio_ffmpeg]
Note: you may need to restart the kernel to use updated packages.


### Environment Overview

- **Action Space**: `Discrete(6)` — The agent can choose from 6 possible actions.
- **Observation Shape**: `(84, 84, 4)` — The input to the agent is a stack of 4 grayscale frames of size 84×84, capturing temporal information for better decision-making.


In [17]:
print("Action space:", env.action_space)
print("Observation shape:", env.observation_space.shape)

Action space: Discrete(6)
Observation shape: (84, 84, 4)


## Baseline DQN

In [22]:
import gymnasium as gym
from stable_baselines3 import DQN
from stable_baselines3.common.vec_env import DummyVecEnv, VecFrameStack
from stable_baselines3.common.env_util import make_atari_env
from stable_baselines3.common.evaluation import evaluate_policy

# Create environment
env = make_atari_env("ALE/DemonAttack-v5", n_envs=1, seed=0)
env = VecFrameStack(env, n_stack=4)

# Hyperparameters 
total_timesteps = 100_000
test_episodes = 20

### DQN Model Training Configuration

A Deep Q-Network (DQN) is initialized with a convolutional neural network (CnnPolicy) to handle image-based input from the Atari environment. Key hyperparameters include:

- **Learning Rate**: `0.0001` — Controls how quickly the model updates.
- **Gamma**: `0.99` — High discount factor for long-term rewards.
- **TensorBoard Logging**: Enabled for monitoring training progress.

The model is trained for `total_timesteps` and saved as `"dqn_demonattack"` for later evaluation or inference.


In [23]:
model = DQN(
    "CnnPolicy",
    env,
    learning_rate=0.0001,
    gamma=0.99,
    verbose=1,
    tensorboard_log="./dqn_demonattack_tensorboard/"
)

model.learn(
    total_timesteps=total_timesteps,
    tb_log_name="dqn_cleanrun"
)
model.save("dqn_demonattack")

Using cpu device
Wrapping the env in a VecTransposeImage.
Logging to ./dqn_demonattack_tensorboard/dqn_cleanrun_1


/Users/anusreemohanan/python 3.11.7/.venv/lib/python3.11/site-packages/stable_baselines3/common/buffers.py:242: UserWarning: This system does not have apparently enough memory to store the complete replay buffer 56.46GB > 1.18GB
  warnings.warn(


----------------------------------
| rollout/            |          |
|    ep_len_mean      | 522      |
|    ep_rew_mean      | 90       |
|    exploration_rate | 0.989    |
| time/               |          |
|    episodes         | 4        |
|    fps              | 293      |
|    time_elapsed     | 0        |
|    total_timesteps  | 114      |
| train/              |          |
|    learning_rate    | 0.0001   |
|    loss             | 0.0148   |
|    n_updates        | 3        |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 484      |
|    ep_rew_mean      | 75       |
|    exploration_rate | 0.98     |
| time/               |          |
|    episodes         | 8        |
|    fps              | 147      |
|    time_elapsed     | 1        |
|    total_timesteps  | 214      |
| train/              |          |
|    learning_rate    | 0.0001   |
|    loss             | 0.0423   |
|    n_updates      

## 📊 DQN Agent Training Results on DemonAttack-v5 (100,000 Steps)

### 🧾 Training Summary
- **Environment**: `DemonAttack-v5`
- **Total Steps**: 100,000
- **Training Time**: ~18.83 minutes
- **TensorBoard Run**: `dqn_cleanrun_1`

---

### 📈 Key Metrics and Trends

1. **Episode Length (`rollout/ep_len_mean`)**  
   - **Trend**: Gradual increase from ~700 to over 1060  
   - **Insight**: The agent is surviving longer, learning to avoid failure effectively.

2. **Episode Reward (`rollout/ep_rew_mean`)**  
   - **Trend**: Steady increase from ~90 to ~260  
   - **Insight**: Higher rewards per episode indicate successful learning.

3. **Exploration Rate (`rollout/exploration_rate`)**  
   - **Trend**: Drops sharply to 0.05 by ~20k steps, then plateaus  
   - **Insight**: The agent transitions from exploration to exploitation—expected with ε-greedy decay.

4. **FPS (`time/fps`)**  
   - **Average**: ~88–92 frames per second  
   - **Insight**: Smooth training on local hardware (MacBook Air) without major slowdowns.

5. **Training Loss (`train/loss`)**  
   - **Trend**: Slight fluctuations between 0.014–0.03  
   - **Insight**: Low and stable loss indicates consistent Q-value u


## 📊 Evaluating Trained DQN Agent

After training the Deep Q-Learning agent on the `DemonAttack-v5` environment, we evaluate its performance over multiple episodes using Stable-Baselines3's `evaluate_policy` utility.

In [25]:
# Evaluate performance
mean_reward, std_reward = evaluate_policy(model, env, n_eval_episodes=test_episodes, render=False)
print(f"Mean reward over {test_episodes} episodes: {mean_reward:.2f} ± {std_reward:.2f}")

Mean reward over 20 episodes: 453.50 ± 323.41


The trained DQN agent achieved a mean reward of 426.25 ± 293.94 over 20 test episodes, indicating strong performance with some variability across runs.

## Inspecting Q-Values from Trained DQN Agent

The following code snippet demonstrates how to extract Q-values from a trained DQN model for a given environment observation in `DemonAttack-v5`.


In [14]:
import torch

# Reset the environment to get the initial observation
obs = env.reset()

# Load your trained model (optional)
from stable_baselines3 import DQN
model = DQN.load("dqn_demonattack")

# Convert obs to tensor and permute to match CNN input
obs_tensor = torch.tensor(obs).float().permute(0, 3, 1, 2).to(model.device)

# Get Q-values
with torch.no_grad():
    q_values = model.q_net(obs_tensor)
    q_values_np = q_values.cpu().numpy()

# Print Q-values and best action
print("📈 Q-values for each action:", q_values_np)
print("🤖 Chosen action:", np.argmax(q_values_np))

📈 Q-values for each action: [[1.3393769 1.3293765 1.2377633 1.1417037 1.2660674 1.1625904]]
🤖 Chosen action: 0


/Users/anusreemohanan/python 3.11.7/.venv/lib/python3.11/site-packages/stable_baselines3/common/buffers.py:242: UserWarning: This system does not have apparently enough memory to store the complete replay buffer 56.46GB > 1.20GB
  warnings.warn(


Action 0 has the highest expected reward, so the agent selects it. This shows how the DQN uses its neural network to pick the best action based on learned Q-values.

## 🎥 Recording Gameplay of Trained DQN Agent

The following script captures video of the trained DQN agent playing **DemonAttack-v5** using the `RecordVideo` wrapper from Gymnasium.

In [26]:
from stable_baselines3 import DQN
from stable_baselines3.common.vec_env import DummyVecEnv, VecFrameStack
from stable_baselines3.common.env_util import make_atari_env
from gymnasium.wrappers import RecordVideo
import os

# Load trained model
model = DQN.load("dqn_demonattack")

# Set up video folder
video_folder = "./videos"
os.makedirs(video_folder, exist_ok=True)

# Create raw Atari env
base_env = make_atari_env("ALE/DemonAttack-v5", n_envs=1, seed=0)
env = base_env.envs[0]  # Get raw gym.Env

# Wrap with RecordVideo before vectorizing
env = RecordVideo(env, video_folder=video_folder, name_prefix="dqn_demo", episode_trigger=lambda ep: True)

# Re-wrap in vectorized and stacked format
vec_env = DummyVecEnv([lambda: env])
vec_env = VecFrameStack(vec_env, n_stack=4)

# Record 3 episodes
for ep in range(3):
    obs = vec_env.reset()
    done = False
    total_reward = 0
    while not done:
        action, _ = model.predict(obs, deterministic=True)
        obs, reward, done, _ = vec_env.step(action)
        total_reward += reward[0]
    print(f"🎮 Episode {ep+1} Reward: {total_reward}")

vec_env.close()


/Users/anusreemohanan/python 3.11.7/.venv/lib/python3.11/site-packages/stable_baselines3/common/buffers.py:242: UserWarning: This system does not have apparently enough memory to store the complete replay buffer 56.46GB > 1.30GB
  warnings.warn(
/Users/anusreemohanan/python 3.11.7/.venv/lib/python3.11/site-packages/gymnasium/wrappers/record_video.py:94: UserWarning: WARN: Overwriting existing videos at /Users/anusreemohanan/videos folder (try specifying a different `video_folder` for the `RecordVideo` wrapper if this is not desired)
  logger.warn(
/Users/anusreemohanan/python 3.11.7/.venv/lib/python3.11/site-packages/gymnasium/utils/passive_env_checker.py:335: UserWarning: WARN: No render fps was declared in the environment (env.metadata['render_fps'] is None or not defined), rendering may occur at inconsistent fps.
  logger.warn(


MoviePy - Building video /Users/anusreemohanan/videos/dqn_demo-episode-0.mp4.
MoviePy - Writing video /Users/anusreemohanan/videos/dqn_demo-episode-0.mp4



MoviePy - Done !
MoviePy - video ready /Users/anusreemohanan/videos/dqn_demo-episode-0.mp4
🎮 Episode 1 Reward: 21.0
MoviePy - Building video /Users/anusreemohanan/videos/dqn_demo-episode-1.mp4.
MoviePy - Writing video /Users/anusreemohanan/videos/dqn_demo-episode-1.mp4



MoviePy - Done !
MoviePy - video ready /Users/anusreemohanan/videos/dqn_demo-episode-1.mp4
🎮 Episode 2 Reward: 1.0


MoviePy - Building video /Users/anusreemohanan/videos/dqn_demo-episode-2.mp4.
MoviePy - Writing video /Users/anusreemohanan/videos/dqn_demo-episode-2.mp4



MoviePy - Done !
MoviePy - video ready /Users/anusreemohanan/videos/dqn_demo-episode-2.mp4
🎮 Episode 3 Reward: 6.0
MoviePy - Building video /Users/anusreemohanan/videos/dqn_demo-episode-3.mp4.
MoviePy - Writing video /Users/anusreemohanan/videos/dqn_demo-episode-3.mp4



MoviePy - Done !
MoviePy - video ready /Users/anusreemohanan/videos/dqn_demo-episode-3.mp4


## Base DQN with Adjustable Hyperparameters (ε-greedy)
This configuration applies an adjustable ε-greedy exploration policy to train a DQN agent on the **DemonAttack-v5** Atari environment.
### 🔧 Hyperparameter Breakdown

| Parameter                 | Value        | Explanation |
|--------------------------|--------------|-------------|
| `learning_rate`          | `0.0005`     | Controls how quickly the model updates Q-values. A slightly higher value than baseline (`0.0001`) speeds up learning but may risk instability. |
| `gamma` (discount factor)| `0.90`       | Determines the importance of future rewards. A lower gamma (compared to `0.99` in baseline) favors immediate rewards and accelerates convergence. |
| `exploration_initial_eps`| `1.0`        | Sets the starting epsilon value. The agent begins by exploring randomly. |
| `exploration_final_eps`  | `0.05`       | Minimum epsilon value. As training progresses, the agent shifts from exploration to exploitation. |
| `exploration_fraction`   | `0.2`        | Indicates that epsilon will decay from 1.0 to 0.05 over the first 20% of training timesteps (20,000 out of 100,000). After that, the agent mostly exploits learned policies. |
| `total_timesteps`        | `100,000`    | Total number of frames the agent interacts with the environment. |
| `tb_log_name`            | `"dqn_epsilon"` | TensorBoard log name to track performance metrics for this configuration. |


In [27]:
from stable_baselines3 import DQN
from stable_baselines3.common.env_util import make_atari_env
from stable_baselines3.common.vec_env import VecFrameStack
from stable_baselines3.common.evaluation import evaluate_policy

# Create and wrap the environment
env = make_atari_env("ALE/DemonAttack-v5", n_envs=1, seed=0)
env = VecFrameStack(env, n_stack=4)

# ε-greedy DQN with adjustable params
model = DQN(
    "CnnPolicy",
    env,
    learning_rate=0.0005,   
    gamma=0.90,            
    exploration_fraction=0.2,     # Portion of training with epsilon decay
    exploration_initial_eps=1.0,  # Starting ε
    exploration_final_eps=0.05,   # Minimum ε
    verbose=1,
    tensorboard_log="./dqn_demonattack_tensorboard/"
)

total_timesteps=100_000
model.learn(
    total_timesteps=total_timesteps,
    tb_log_name="dqn_epsilon"
)
model.save("dqn_epsilon")

Using cpu device
Wrapping the env in a VecTransposeImage.
Logging to ./dqn_demonattack_tensorboard/dqn_epsilon_1


/Users/anusreemohanan/python 3.11.7/.venv/lib/python3.11/site-packages/stable_baselines3/common/buffers.py:242: UserWarning: This system does not have apparently enough memory to store the complete replay buffer 56.46GB > 1.19GB
  warnings.warn(


----------------------------------
| rollout/            |          |
|    ep_len_mean      | 522      |
|    ep_rew_mean      | 90       |
|    exploration_rate | 0.995    |
| time/               |          |
|    episodes         | 4        |
|    fps              | 263      |
|    time_elapsed     | 0        |
|    total_timesteps  | 114      |
| train/              |          |
|    learning_rate    | 0.0005   |
|    loss             | 0.0152   |
|    n_updates        | 3        |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 375      |
|    ep_rew_mean      | 55       |
|    exploration_rate | 0.992    |
| time/               |          |
|    episodes         | 8        |
|    fps              | 177      |
|    time_elapsed     | 0        |
|    total_timesteps  | 160      |
| train/              |          |
|    learning_rate    | 0.0005   |
|    loss             | 0.0187   |
|    n_updates      

## Training Details Summary

### Timesteps
- **Total Training Steps**: `100,000`

---

### Learning Rates
| Model            | Learning Rate |
|------------------|---------------|
| `dqn_cleanrun_1` | `0.0001`       |
| `dqn_epsilon_1`  | `0.0005`       |

---

### Exploration Strategy
| Model            | Strategy           |
|------------------|--------------------|
| `dqn_cleanrun_1` | Default             |
| `dqn_epsilon_1`  | Custom ε-greedy     |

---

### Performance Insight

The `dqn_epsilon_1` configuration demonstrated effective learning progression under a modified ε-greedy strategy. Despite slightly noisier loss values, the agent achieved:
- A **mean reward of ~426**
- **Stable episode lengths**

This suggests that the tuned exploration settings allowed the agent to explore efficiently while still converging to a solid policy.


## Model Evaluation: `dqn_epsilon`

This code evaluates the performance of the trained Deep Q-Learning agent on the `DemonAttack-v5` Atari environment using Stable-Baselines3’s built-in `evaluate_policy` function.


In [29]:
from stable_baselines3.common.evaluation import evaluate_policy

# Reload the environment and model
eval_env = make_atari_env("ALE/DemonAttack-v5", n_envs=1, seed=42)
eval_env = VecFrameStack(eval_env, n_stack=4)

# Load the trained model
model = DQN.load("dqn_epsilon")

# Set number of evaluation episodes
test_episodes = 20

# Evaluate the model
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=test_episodes, render=False)
print(f"Mean reward over {test_episodes} episodes: {mean_reward:.2f} ± {std_reward:.2f}")


/Users/anusreemohanan/python 3.11.7/.venv/lib/python3.11/site-packages/stable_baselines3/common/buffers.py:242: UserWarning: This system does not have apparently enough memory to store the complete replay buffer 56.46GB > 1.20GB
  warnings.warn(


Mean reward over 20 episodes: 220.75 ± 78.54


The ε-greedy DQN model (dqn_epsilon) achieved a mean reward of 220.75 ± 78.54 over 20 evaluation episodes on the DemonAttack-v5 environment. This indicates that the agent has learned to perform moderately well, consistently achieving decent scores, though the standard deviation suggests some variability in performance across episodes. Further tuning or exploration strategy adjustments may help improve stability and peak performance.

##  QRDQN Agent with NoisyNet for DemonAttack-v5

This setup uses `QRDQN` (Quantile Regression DQN) from `sb3_contrib` along with NoisyNet to improve exploration in the `DemonAttack-v5` Atari environment. NoisyNet replaces the ε-greedy strategy by injecting noise directly into the network weights.


In [1]:
%pip install sb3-contrib

Note: you may need to restart the kernel to use updated packages.


In [1]:
from sb3_contrib import QRDQN

## QRDQN Agent with NoisyNet for DemonAttack-v5

This configuration uses **QRDQN** (Quantile Regression DQN) with **NoisyNet** to train on the **DemonAttack-v5** Atari environment. Instead of ε-greedy exploration, NoisyNet injects learnable noise into network weights to drive exploration.

### 🔧 Hyperparameter Breakdown

| Parameter                   | Value        | Explanation |
|----------------------------|--------------|-------------|
| `learning_rate`            | `0.0005`     | Enables faster Q-value updates. Higher than the baseline DQN rate (`0.0001`) to encourage more responsive learning. |
| `gamma` (discount factor)  | `0.99`       | Emphasizes long-term rewards, similar to the baseline configuration. |
| `exploration_initial_eps` | `0.0`        | No random exploration via epsilon. NoisyNet handles all exploration implicitly. |
| `exploration_final_eps`   | `0.0`        | Remains zero since ε-greedy is disabled. |
| `exploration_strategy`    | `NoisyNet`   | Exploration is achieved by adding noise to the neural network’s weights during training. |
| `total_timesteps`         | `100,000`    | Total environment interaction steps used for training. |
| `tb_log_name`             | `"qrdqn_noisynet"` | TensorBoard run name for monitoring training progress and reward trends. |

---

**Why NoisyNet?**  
Unlike ε-greedy policies that decay over time, NoisyNet facilitates **state-dependent exploration**. The injected noise in linear layers helps the agent explore intelligently without needing manually tuned decay schedules.

**Result Summary:**  
This agent showed competitive performance and faster reward growth during early episodes compared to ε-greedy DQN, though final reward stability may vary depending on noise variance.



In [6]:
from sb3_contrib import QRDQN
from stable_baselines3.common.env_util import make_atari_env
from stable_baselines3.common.vec_env import VecFrameStack

# Env setup
env = make_atari_env("ALE/DemonAttack-v5", n_envs=1, seed=0)
env = VecFrameStack(env, n_stack=4)

# QRDQN with NoisyNet (ε-greedy disabled)
model_noisy = QRDQN(
    "CnnPolicy",
    env,
    learning_rate=0.0005,
    gamma=0.99,
    exploration_initial_eps=0.0,
    exploration_final_eps=0.0,
    verbose=1,
    tensorboard_log="./dqn_demonattack_tensorboard/"
)

# Train for full steps
model_noisy.learn(
    total_timesteps=100_000,
    tb_log_name="qrdqn_noisynet"
)
model_noisy.save("qrdqn_noisynet")


Using cpu device
Wrapping the env in a VecTransposeImage.
Logging to ./dqn_demonattack_tensorboard/qrdqn_noisynet_1


/Users/anusreemohanan/python 3.11.7/.venv/lib/python3.11/site-packages/stable_baselines3/common/buffers.py:242: UserWarning: This system does not have apparently enough memory to store the complete replay buffer 56.46GB > 1.35GB
  warnings.warn(


----------------------------------
| rollout/            |          |
|    exploration_rate | 0        |
| time/               |          |
|    episodes         | 4        |
|    fps              | 95       |
|    time_elapsed     | 4        |
|    total_timesteps  | 431      |
| train/              |          |
|    learning_rate    | 0.0005   |
|    loss             | 1.39     |
|    n_updates        | 82       |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 2.11e+03 |
|    ep_rew_mean      | 400      |
|    exploration_rate | 0        |
| time/               |          |
|    episodes         | 8        |
|    fps              | 73       |
|    time_elapsed     | 27       |
|    total_timesteps  | 2052     |
| train/              |          |
|    learning_rate    | 0.0005   |
|    loss             | 0.879    |
|    n_updates        | 487      |
----------------------------------
--------------------

## QRDQN Agent with NoisyNet for DemonAttack-v5

This experiment implements a **QRDQN (Quantile Regression DQN)** agent with **NoisyNet exploration** on the Atari environment **DemonAttack-v5**. Unlike ε-greedy exploration, NoisyNet introduces trainable noise in the network weights to drive exploration, making epsilon scheduling unnecessary.

---

### Configuration Summary

| Hyperparameter              | Value         | Description                                                                 |
|----------------------------|---------------|-----------------------------------------------------------------------------|
| `Algorithm`                | `QRDQN`       | A distributional RL variant of DQN                                          |
| `Exploration Strategy`     | `NoisyNet`    | No need for ε-greedy; exploration via parameter noise                       |
| `learning_rate`            | `0.0005`      | Learning rate for optimizer                                                 |
| `gamma` (discount factor)  | `0.99`        | Importance of future rewards                                                |
| `exploration_initial_eps`  | `0.0`         | Not used (NoisyNet handles exploration)                                     |
| `exploration_final_eps`    | `0.0`         | Constant 0 (no ε-greedy decay)                                              |
| `total_timesteps`          | `100,000`     | Total training steps                                                        |
| `tb_log_name`              | `qrdqn_noisynet` | TensorBoard log label                                                      |

---

### TensorBoard Observations

#### **Rollout Performance**

| Metric              | Final Value     | Insights                                                  |
|---------------------|------------------|-----------------------------------------------------------|
| `ep_len_mean`       | ~1013 steps      | Long-lasting episodes (agent survives longer)             |
| `ep_rew_mean`       | ~327 points      | Highest reward among all models                           |
| `exploration_rate`  | 0                | Remained 0 due to NoisyNet (as expected)                  |

#### **Training Performance**

| Metric               | Value           | Comment                                                   |
|----------------------|------------------|-----------------------------------------------------------|
| `learning_rate`      | 0.0005           | Fixed throughout                                          |
| `train/loss`         | Peaks ~4.5       | Higher due to QRDQN's distributional output               |
| `n_updates`          | ~24,971          | Update count shows stable training                        |
| `fps` (speed)        | ~73 FPS          | Slightly slower due to QRDQN complexity                   |

---

### Evaluation Sample

```text
ep_len_mean      | 2110
ep_rew_mean      | 400
exploration_rate | 0
episodes         | 8
fps              | 73
time_elapsed     | 27 seconds
total_timesteps  | 2052
learning_rate    | 0.0005
loss             | 4.5
n_updates        | 24,971


## Evaluation: QRDQN with NoisyNet

After training the QRDQN agent with NoisyNet exploration, we evaluate its performance across 20 episodes using the `evaluate_policy` utility from Stable-Baselines3.

In [7]:
from stable_baselines3.common.evaluation import evaluate_policy

# Evaluate NoisyNet model
mean_reward, std_reward = evaluate_policy(model_noisy, env, n_eval_episodes=20)
print(f"🎯 NoisyNet QRDQN Mean reward over 20 episodes: {mean_reward:.2f} ± {std_reward:.2f}")

🎯 NoisyNet QRDQN Mean reward over 20 episodes: 423.75 ± 288.37


The NoisyNet QRDQN agent achieved a high average reward of 423.75 over 20 episodes, demonstrating strong performance and effective policy learning, though the high standard deviation (±288.37) reflects variability in episode outcomes.

MIT License

Copyright (c) 2025 Anusree Mohanan

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.